File: container-nix/bb-sec_go-api/src/routes/health.go
package routes

import (
	"encoding/json"
	"net/http"
	"sync"

	"github.com/diegonmarcos/go-api/config"
	"github.com/diegonmarcos/go-api/services"
	"github.com/go-chi/chi/v5"
)

// RegisterHealth registers all tiered health endpoints under /go/health/*.
func RegisterHealth(r chi.Router, cfg *config.AppConfig) {
	r.Get("/go/health", healthAlive)
	r.Get("/go/health/declared", healthDeclared(cfg))
	r.Get("/go/health/deployed", healthDeployed(cfg))
	r.Get("/go/health/deployed/{vm_id}", healthDeployedVM(cfg))
	r.Get("/go/health/drift", healthDrift(cfg))
	r.Get("/go/health/status", healthStatus(cfg))
	r.Get("/go/health/status/{vm_id}", healthStatusVM(cfg))
}

func writeJSON(w http.ResponseWriter, v interface{}) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

func writeError(w http.ResponseWriter, code int, msg string) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	json.NewEncoder(w).Encode(map[string]string{"error": msg})
}

// Tier 0: alive
func healthAlive(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, map[string]string{"status": "ok"})
}

// Tier 0: declared (no SSH)
func healthDeclared(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		vms := map[string]interface{}{}
		totalServices := 0
		totalContainers := 0

		for vmID, vmMap := range cfg.AllVmServices {
			svcMap := map[string][]string{}
			for svcName, containers := range vmMap.Services {
				totalServices++
				totalContainers += len(containers)
				svcMap[svcName] = containers
			}
			vms[vmID] = map[string]interface{}{
				"label":    vmMap.Label,
				"services": svcMap,
			}
		}

		writeJSON(w, map[string]interface{}{
			"vms":     vms,
			"domains": cfg.ContainerDomainMap,
			"totals": map[string]int{
				"vms":        len(cfg.AllVmServices),
				"services":   totalServices,
				"containers": totalContainers,
			},
		})
	}
}

// Tier 1: deployed (docker ps -a via SSH, parallel)
func healthDeployed(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		type vmResult struct {
			vmID       string
			label      string
			containers []services.ContainerStatus
		}

		var mu sync.Mutex
		var wg sync.WaitGroup
		results := make([]vmResult, 0, len(cfg.AllVmServices))

		for vmID, vmMap := range cfg.AllVmServices {
			wg.Add(1)
			go func(id, label string) {
				defer wg.Done()
				sshCfg, ok := cfg.VmSSH[id]
				if !ok {
					return
				}
				containers := services.BatchContainerStatuses(sshCfg)
				mu.Lock()
				results = append(results, vmResult{id, label, containers})
				mu.Unlock()
			}(vmID, vmMap.Label)
		}
		wg.Wait()

		vms := map[string]interface{}{}
		globalRunning, globalStopped, globalTotal := 0, 0, 0

		for _, res := range results {
			running, stopped := 0, 0
			details := make([]map[string]string, 0, len(res.containers))
			for _, c := range res.containers {
				if c.State == "running" {
					running++
				} else {
					stopped++
				}
				details = append(details, map[string]string{
					"name": c.Name, "state": c.State, "ports": c.Ports,
				})
			}
			globalRunning += running
			globalStopped += stopped
			globalTotal += len(res.containers)
			vms[res.vmID] = map[string]interface{}{
				"label": res.label, "containers": details,
				"running": running, "stopped": stopped, "total": len(res.containers),
			}
		}

		writeJSON(w, map[string]interface{}{
			"vms": vms,
			"summary": map[string]int{
				"running": globalRunning, "stopped": globalStopped, "total": globalTotal,
			},
		})
	}
}

// Tier 1: deployed per-VM
func healthDeployedVM(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		vmID := chi.URLParam(r, "vm_id")
		vmMap, ok := cfg.AllVmServices[vmID]
		if !ok {
			writeError(w, 404, "Unknown VM: "+vmID)
			return
		}
		sshCfg, ok := cfg.VmSSH[vmID]
		if !ok {
			writeError(w, 500, "No SSH config for VM: "+vmID)
			return
		}

		containers := services.BatchContainerStatuses(sshCfg)
		running, stopped := 0, 0
		details := make([]map[string]string, 0, len(containers))
		for _, c := range containers {
			if c.State == "running" {
				running++
			} else {
				stopped++
			}
			details = append(details, map[string]string{
				"name": c.Name, "state": c.State, "ports": c.Ports,
			})
		}

		writeJSON(w, map[string]interface{}{
			"vm_id": vmID, "label": vmMap.Label,
			"containers": details,
			"running": running, "stopped": stopped, "total": len(containers),
		})
	}
}

// Tier 2: drift
func healthDrift(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		type vmResult struct {
			vmID       string
			containers []services.ContainerStatus
		}

		var mu sync.Mutex
		var wg sync.WaitGroup
		deployedMap := map[string][]services.ContainerStatus{}

		for vmID := range cfg.AllVmServices {
			wg.Add(1)
			go func(id string) {
				defer wg.Done()
				sshCfg, ok := cfg.VmSSH[id]
				if !ok {
					return
				}
				containers := services.BatchContainerStatuses(sshCfg)
				mu.Lock()
				deployedMap[id] = containers
				mu.Unlock()
			}(vmID)
		}
		wg.Wait()

		vms := map[string]interface{}{}
		sumDeclared, sumDeployed, sumMissing, sumExtra := 0, 0, 0, 0

		for vmID, vmMap := range cfg.AllVmServices {
			declared := []string{}
			for _, containers := range vmMap.Services {
				declared = append(declared, containers...)
			}

			deployedRaw := deployedMap[vmID]
			deployed := make([]string, 0, len(deployedRaw))
			for _, c := range deployedRaw {
				deployed = append(deployed, c.Name)
			}

			missing := diff(declared, deployed)
			extra := diff(deployed, declared)

			sumDeclared += len(declared)
			sumDeployed += len(deployed)
			sumMissing += len(missing)
			sumExtra += len(extra)

			vms[vmID] = map[string]interface{}{
				"label": vmMap.Label, "declared": declared, "deployed": deployed,
				"missing": missing, "extra": extra,
			}
		}

		writeJSON(w, map[string]interface{}{
			"vms": vms,
			"summary": map[string]interface{}{
				"declared": sumDeclared, "deployed": sumDeployed,
				"missing": sumMissing, "extra": sumExtra,
				"drift": sumMissing > 0 || sumExtra > 0,
			},
		})
	}
}

// Tier 3: status (comprehensive)
func healthStatus(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		type vmResult struct {
			vmID       string
			label      string
			sshOK      bool
			pingOK     bool
			containers []services.ContainerStatus
		}

		var mu sync.Mutex
		var wg sync.WaitGroup
		results := map[string]vmResult{}

		for vmID, vmMap := range cfg.AllVmServices {
			wg.Add(1)
			go func(id, label string) {
				defer wg.Done()
				sshCfg, ok := cfg.VmSSH[id]
				if !ok {
					return
				}
				pingOK := services.CheckPing(sshCfg.Host)
				sshOK := services.CheckSSH(sshCfg)
				var containers []services.ContainerStatus
				if sshOK {
					containers = services.BatchContainerStatuses(sshCfg)
				}
				mu.Lock()
				results[id] = vmResult{id, label, sshOK, pingOK, containers}
				mu.Unlock()
			}(vmID, vmMap.Label)
		}
		wg.Wait()

		vms := map[string]interface{}{}
		for vmID, res := range results {
			health := "unknown"
			if res.sshOK {
				health = "online"
			} else if res.pingOK {
				health = "degraded"
			} else {
				health = "offline"
			}

			running, total := 0, 0
			details := make([]map[string]string, 0, len(res.containers))
			for _, c := range res.containers {
				total++
				if c.State == "running" {
					running++
				}
				details = append(details, map[string]string{
					"name": c.Name, "state": c.State, "ports": c.Ports,
				})
			}

			vms[vmID] = map[string]interface{}{
				"label": res.label, "ssh": res.sshOK, "ping": res.pingOK,
				"health": health, "containers": details,
				"summary": map[string]int{
					"containers_running": running, "containers_total": total,
				},
			}
		}

		writeJSON(w, map[string]interface{}{"vms": vms})
	}
}

// Tier 3: status per-VM
func healthStatusVM(cfg *config.AppConfig) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		vmID := chi.URLParam(r, "vm_id")
		vmMap, ok := cfg.AllVmServices[vmID]
		if !ok {
			writeError(w, 404, "Unknown VM: "+vmID)
			return
		}
		sshCfg, ok := cfg.VmSSH[vmID]
		if !ok {
			writeError(w, 500, "No SSH config for VM: "+vmID)
			return
		}

		pingOK := services.CheckPing(sshCfg.Host)
		sshOK := services.CheckSSH(sshCfg)
		var containers []services.ContainerStatus
		if sshOK {
			containers = services.BatchContainerStatuses(sshCfg)
		}

		health := "unknown"
		if sshOK {
			health = "online"
		} else if pingOK {
			health = "degraded"
		} else {
			health = "offline"
		}

		running, total := 0, 0
		details := make([]map[string]string, 0, len(containers))
		for _, c := range containers {
			total++
			if c.State == "running" {
				running++
			}
			details = append(details, map[string]string{
				"name": c.Name, "state": c.State, "ports": c.Ports,
			})
		}

		writeJSON(w, map[string]interface{}{
			"vm_id": vmID, "label": vmMap.Label,
			"ssh": sshOK, "ping": pingOK, "health": health,
			"containers": details,
			"summary": map[string]int{
				"containers_running": running, "containers_total": total,
			},
		})
	}
}

// diff returns elements in a that are not in b.
func diff(a, b []string) []string {
	set := map[string]bool{}
	for _, v := range b {
		set[v] = true
	}
	var result []string
	for _, v := range a {
		if !set[v] {
			result = append(result, v)
		}
	}
	if result == nil {
		return []string{}
	}
	return result
}

File: bb-sec_rust-api/src/src/src/routes/health.rs
use axum::{
    extract::{Path, State},
    routing::get,
    Json, Router,
};
use serde_json::{json, Value};
use std::sync::Arc;

use crate::error::AppError;
use crate::models::vm::validate_vm_id;
use crate::AppState;

use super::ondemand::{
    batch_container_statuses, compute_service_status, gather_vm_resources, probe_domain_chain,
    probe_vm, round1,
};

pub fn routes() -> Router<Arc<AppState>> {
    Router::new()
        .route("/api/health", get(health_alive))
        .route("/api/health/declared", get(health_declared))
        .route("/api/health/deployed", get(health_deployed))
        .route("/api/health/deployed/{vm_id}", get(health_deployed_vm))
        .route("/api/health/drift", get(health_drift))
        .route("/api/health/status", get(health_status))
        .route("/api/health/status/{vm_id}", get(health_status_vm))
}

// ---------------------------------------------------------------------------
// Tier 0: Alive
// ---------------------------------------------------------------------------

#[utoipa::path(
    get,
    path = "/api/health",
    tag = "Health",
    responses(
        (status = 200, description = "API alive check", body = Value),
    )
)]
pub async fn health_alive() -> Json<Value> {
    Json(json!({ "status": "ok" }))
}

// ---------------------------------------------------------------------------
// Tier 0: Declared (instant, no SSH)
// ---------------------------------------------------------------------------

#[utoipa::path(
    get,
    path = "/api/health/declared",
    tag = "Health",
    responses(
        (status = 200, description = "Declared services and containers from config (no SSH)", body = Value),
    )
)]
pub async fn health_declared(
    State(state): State<Arc<AppState>>,
) -> Json<Value> {
    let mut vms = serde_json::Map::new();
    let mut total_services = 0u32;
    let mut total_containers = 0u32;

    for (vm_id, vm_map) in &state.config.all_vm_services {
        let mut services_map = serde_json::Map::new();
        for (svc_name, containers) in &vm_map.services {
            total_services += 1;
            total_containers += containers.len() as u32;
            services_map.insert(svc_name.clone(), json!(containers));
        }
        vms.insert(vm_id.clone(), json!({
            "label": vm_map.label,
            "services": services_map,
        }));
    }

    let domains: serde_json::Map<String, Value> = state.config.container_domain_map
        .iter()
        .map(|(k, v)| (k.clone(), json!(v)))
        .collect();

    Json(json!({
        "vms": vms,
        "domains": domains,
        "totals": {
            "vms": state.config.all_vm_services.len(),
            "services": total_services,
            "containers": total_containers,
        }
    }))
}

// ---------------------------------------------------------------------------
// Tier 1: Deployed (docker ps -a via SSH)
// ---------------------------------------------------------------------------

#[utoipa::path(
    get,
    path = "/api/health/deployed",
    tag = "Health",
    responses(
        (status = 200, description = "Docker container states across all VMs (parallel SSH)", body = Value),
        (status = 500, description = "Internal error")
    )
)]
pub async fn health_deployed(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, AppError> {
    let mut set = tokio::task::JoinSet::new();

    for (vm_id, vm_map) in &state.config.all_vm_services {
        let vm_id = vm_id.clone();
        let label = vm_map.label.clone();
        let state = Arc::clone(&state);
        set.spawn(async move {
            let containers = batch_container_statuses(&state, &vm_id).await;
            (vm_id, label, containers)
        });
    }

    let mut vms = serde_json::Map::new();
    let mut global_running = 0u32;
    let mut global_stopped = 0u32;
    let mut global_total = 0u32;

    while let Some(result) = set.join_next().await {
        let (vm_id, label, containers) = match result {
            Ok(v) => v,
            Err(_) => continue,
        };

        let mut running = 0u32;
        let mut stopped = 0u32;
        let details: Vec<Value> = containers.iter().map(|(name, st, ports)| {
            if st == "running" {
                running += 1;
            } else {
                stopped += 1;
            }
            json!({ "name": name, "state": st, "ports": ports })
        }).collect();
        let total = details.len() as u32;
        global_running += running;
        global_stopped += stopped;
        global_total += total;

        vms.insert(vm_id, json!({
            "label": label,
            "containers": details,
            "running": running,
            "stopped": stopped,
            "total": total,
        }));
    }

    Ok(Json(json!({
        "vms": vms,
        "summary": {
            "running": global_running,
            "stopped": global_stopped,
            "total": global_total,
        }
    })))
}

#[utoipa::path(
    get,
    path = "/api/health/deployed/{vm_id}",
    tag = "Health",
    params(("vm_id" = String, Path, description = "VM identifier")),
    responses(
        (status = 200, description = "Docker container states for one VM", body = Value),
        (status = 404, description = "Unknown VM"),
    )
)]
pub async fn health_deployed_vm(
    State(state): State<Arc<AppState>>,
    Path(vm_id): Path<String>,
) -> Result<Json<Value>, AppError> {
    if !validate_vm_id(&vm_id) {
        return Err(AppError::bad_request("Invalid VM ID"));
    }
    let vm_map = state.config.all_vm_services.get(&vm_id)
        .ok_or_else(|| AppError::not_found(format!("Unknown VM: {vm_id}")))?;

    let containers = batch_container_statuses(&state, &vm_id).await;
    let mut running = 0u32;
    let mut stopped = 0u32;
    let details: Vec<Value> = containers.iter().map(|(name, st, ports)| {
        if st == "running" {
            running += 1;
        } else {
            stopped += 1;
        }
        json!({ "name": name, "state": st, "ports": ports })
    }).collect();

    Ok(Json(json!({
        "vm_id": vm_id,
        "label": vm_map.label,
        "containers": details,
        "running": running,
        "stopped": stopped,
        "total": details.len(),
    })))
}

// ---------------------------------------------------------------------------
// Tier 2: Drift (declared vs deployed)
// ---------------------------------------------------------------------------

#[utoipa::path(
    get,
    path = "/api/health/drift",
    tag = "Health",
    responses(
        (status = 200, description = "Declared vs deployed container drift analysis", body = Value),
        (status = 500, description = "Internal error")
    )
)]
pub async fn health_drift(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, AppError> {
    let mut set = tokio::task::JoinSet::new();

    for (vm_id, _vm_map) in &state.config.all_vm_services {
        let vm_id = vm_id.clone();
        let state = Arc::clone(&state);
        set.spawn(async move {
            let containers = batch_container_statuses(&state, &vm_id).await;
            (vm_id, containers)
        });
    }

    let mut deployed_map: std::collections::HashMap<String, Vec<(String, String, String)>> =
        std::collections::HashMap::new();
    while let Some(result) = set.join_next().await {
        if let Ok((vm_id, containers)) = result {
            deployed_map.insert(vm_id, containers);
        }
    }

    let mut vms = serde_json::Map::new();
    let mut summary_declared = 0u32;
    let mut summary_deployed = 0u32;
    let mut summary_missing = 0u32;
    let mut summary_extra = 0u32;

    for (vm_id, vm_map) in &state.config.all_vm_services {
        let declared: Vec<String> = vm_map.services
            .values()
            .flat_map(|c| c.iter().cloned())
            .collect();

        let deployed_raw = deployed_map.get(vm_id).cloned().unwrap_or_default();
        let deployed: Vec<String> = deployed_raw.iter().map(|(n, _, _)| n.clone()).collect();

        let missing: Vec<&String> = declared.iter()
            .filter(|d| !deployed.contains(d))
            .collect();

        let extra: Vec<&String> = deployed.iter()
            .filter(|d| !declared.contains(d))
            .collect();

        summary_declared += declared.len() as u32;
        summary_deployed += deployed.len() as u32;
        summary_missing += missing.len() as u32;
        summary_extra += extra.len() as u32;

        vms.insert(vm_id.clone(), json!({
            "label": vm_map.label,
            "declared": declared,
            "deployed": deployed,
            "missing": missing,
            "extra": extra,
        }));
    }

    let drift = summary_missing > 0 || summary_extra > 0;

    Ok(Json(json!({
        "vms": vms,
        "summary": {
            "declared": summary_declared,
            "deployed": summary_deployed,
            "missing": summary_missing,
            "extra": summary_extra,
            "drift": drift,
        }
    })))
}

// ---------------------------------------------------------------------------
// Tier 3: Status (comprehensive — provider + SSH + resources + containers + routes)
// ---------------------------------------------------------------------------

#[utoipa::path(
    get,
    path = "/api/health/status",
    tag = "Health",
    responses(
        (status = 200, description = "Comprehensive health: provider state, SSH, resources, containers, route probes", body = Value),
        (status = 500, description = "Internal error")
    )
)]
pub async fn health_status(
    State(state): State<Arc<AppState>>,
) -> Result<Json<Value>, AppError> {
    // Parallel: probe all VMs + gather resources + probe routes
    let mut vm_set = tokio::task::JoinSet::new();
    let mut res_set = tokio::task::JoinSet::new();

    for (vm_id, vm_map) in &state.config.all_vm_services {
        let vm_id_1 = vm_id.clone();
        let vm_id_2 = vm_id.clone();
        let label = vm_map.label.clone();
        let services = vm_map.services.clone();
        let s1 = Arc::clone(&state);
        let s2 = Arc::clone(&state);

        vm_set.spawn(async move {
            let probe = probe_vm(&s1, &vm_id_1).await;
            (vm_id_1, label, services, probe)
        });
        res_set.spawn(async move {
            let resources = gather_vm_resources(&s2, &vm_id_2).await;
            (vm_id_2, resources)
        });
    }

    // Route probes in parallel
    let mut route_set = tokio::task::JoinSet::new();
    let bearer = state.config.authelia_bearer_token.clone();
    for entry in &state.config.route_check_domains {
        let http = state.http.clone();
        let domain = entry.domain.clone();
        let service = entry.service.clone();
        let token = bearer.clone();
        route_set.spawn(async move {
            probe_domain_chain(&http, &domain, &service, token.as_deref()).await
        });
    }

    // OCI Object Storage
    let storage_state = Arc::clone(&state);
    let storage_handle = tokio::spawn(async move {
        let quota = storage_state.config.oci_storage_quota_gb;
        crate::services::oci::get_object_storage_info(&storage_state.http, &storage_state.config, quota).await
    });

    // Collect VM probes
    let mut vm_data: std::collections::HashMap<String, (String, std::collections::HashMap<String, Vec<String>>, super::ondemand::VmProbe)> =
        std::collections::HashMap::new();
    while let Some(result) = vm_set.join_next().await {
        if let Ok((vm_id, label, services, probe)) = result {
            vm_data.insert(vm_id, (label, services, probe));
        }
    }

    // Collect resources
    let mut resources_map: std::collections::HashMap<String, Value> = std::collections::HashMap::new();
    while let Some(result) = res_set.join_next().await {
        if let Ok((vm_id, resources)) = result {
            resources_map.insert(vm_id, resources);
        }
    }

    // Build VM responses
    let mut vms = serde_json::Map::new();
    let mut global_services_up = 0u32;
    let mut global_services_down = 0u32;
    let mut global_services_partial = 0u32;
    let mut global_containers_running = 0u32;
    let mut global_containers_total = 0u32;
    let mut total_cores = 0u32;
    let mut total_ram_mb = 0u64;
    let mut total_disk_gb = 0.0f64;
    let mut sum_cpu_pct = 0.0f64;
    let mut sum_mem_pct = 0.0f64;
    let mut vm_count = 0u32;

    for (vm_id, (label, services, probe)) in &vm_data {
        let mut svc_map = serde_json::Map::new();
        let mut vm_services_up = 0u32;
        let mut vm_services_down = 0u32;
        let mut vm_services_partial = 0u32;
        let mut vm_containers_running = 0u32;
        let mut vm_containers_total = 0u32;

        for (svc_name, containers) in services {
            let svc_status = compute_service_status(containers, &probe.container_data);
            let status_str = svc_status["status"].as_str().unwrap_or("down");
            match status_str {
                "up" => vm_services_up += 1,
                "partial" => vm_services_partial += 1,
                _ => vm_services_down += 1,
            }
            if let Some(ctrs) = svc_status["containers"].as_array() {
                for c in ctrs {
                    vm_containers_total += 1;
                    if c["state"].as_str() == Some("running") {
                        vm_containers_running += 1;
                    }
                }
            }
            svc_map.insert(svc_name.clone(), svc_status);
        }

        global_services_up += vm_services_up;
        global_services_down += vm_services_down;
        global_services_partial += vm_services_partial;
        global_containers_running += vm_containers_running;
        global_containers_total += vm_containers_total;

        let resources = resources_map.get(vm_id).cloned().unwrap_or(json!(null));
        if resources.get("error").is_none() && !resources.is_null() {
            total_cores += resources["resources"]["cpu"]["cores"].as_u64().unwrap_or(0) as u32;
            total_ram_mb += resources["resources"]["memory"]["total_mb"].as_u64().unwrap_or(0);
            total_disk_gb += resources["resources"]["disk"]["total_gb"].as_f64().unwrap_or(0.0);
            sum_cpu_pct += resources["resources"]["cpu"]["usage_pct"].as_f64().unwrap_or(0.0);
            sum_mem_pct += resources["resources"]["memory"]["usage_pct"].as_f64().unwrap_or(0.0);
            vm_count += 1;
        }

        vms.insert(vm_id.clone(), json!({
            "label": label,
            "provider_state": probe.provider_state,
            "ssh": probe.ssh,
            "ping": probe.ping,
            "health": probe.health,
            "services": svc_map,
            "resources": resources,
            "summary": {
                "services_up": vm_services_up,
                "services_down": vm_services_down,
                "services_partial": vm_services_partial,
                "containers_running": vm_containers_running,
                "containers_total": vm_containers_total,
            }
        }));
    }

    // Collect route probes
    let mut routes = vec![];
    let mut routes_healthy = 0u32;
    let mut routes_total = 0u32;
    while let Some(result) = route_set.join_next().await {
        if let Ok(val) = result {
            routes_total += 1;
            if val["healthy"].as_bool() == Some(true) {
                routes_healthy += 1;
            }
            routes.push(val);
        }
    }
    routes.sort_by(|a, b| {
        a["service"].as_str().unwrap_or("").cmp(b["service"].as_str().unwrap_or(""))
    });

    // Collect object storage
    let object_storage = match storage_handle.await {
        Ok(Ok(data)) => data,
        Ok(Err(e)) => json!({"error": e}),
        Err(_) => json!({"error": "task failed"}),
    };

    let avg_cpu_pct = if vm_count > 0 { sum_cpu_pct / vm_count as f64 } else { 0.0 };
    let avg_mem_pct = if vm_count > 0 { sum_mem_pct / vm_count as f64 } else { 0.0 };

    Ok(Json(json!({
        "vms": vms,
        "routes": {
            "probes": routes,
            "healthy": routes_healthy,
            "total": routes_total,
        },
        "object_storage": object_storage,
        "summary": {
            "services_up": global_services_up,
            "services_down": global_services_down,
            "services_partial": global_services_partial,
            "containers_running": global_containers_running,
            "containers_total": global_containers_total,
            "total_cpu_cores": total_cores,
            "total_ram_mb": total_ram_mb,
            "total_disk_gb": round1(total_disk_gb),
            "avg_cpu_pct": round1(avg_cpu_pct),
            "avg_mem_pct": round1(avg_mem_pct),
        }
    })))
}

#[utoipa::path(
    get,
    path = "/api/health/status/{vm_id}",
    tag = "Health",
    params(("vm_id" = String, Path, description = "VM identifier")),
    responses(
        (status = 200, description = "Comprehensive per-VM health status", body = Value),
        (status = 404, description = "Unknown VM"),
    )
)]
pub async fn health_status_vm(
    State(state): State<Arc<AppState>>,
    Path(vm_id): Path<String>,
) -> Result<Json<Value>, AppError> {
    if !validate_vm_id(&vm_id) {
        return Err(AppError::bad_request("Invalid VM ID"));
    }
    let vm_map = state.config.all_vm_services.get(&vm_id)
        .ok_or_else(|| AppError::not_found(format!("Unknown VM: {vm_id}")))?;

    let label = vm_map.label.clone();
    let services = vm_map.services.clone();

    // Parallel: probe + resources
    let s1 = Arc::clone(&state);
    let s2 = Arc::clone(&state);
    let vm_id_1 = vm_id.clone();
    let vm_id_2 = vm_id.clone();

    let (probe, resources) = tokio::join!(
        async { probe_vm(&s1, &vm_id_1).await },
        async { gather_vm_resources(&s2, &vm_id_2).await },
    );

    let mut svc_map = serde_json::Map::new();
    let mut services_up = 0u32;
    let mut services_down = 0u32;
    let mut services_partial = 0u32;
    let mut containers_running = 0u32;
    let mut containers_total = 0u32;

    for (svc_name, containers) in &services {
        let svc_status = compute_service_status(containers, &probe.container_data);
        let status_str = svc_status["status"].as_str().unwrap_or("down");
        match status_str {
            "up" => services_up += 1,
            "partial" => services_partial += 1,
            _ => services_down += 1,
        }
        if let Some(ctrs) = svc_status["containers"].as_array() {
            for c in ctrs {
                containers_total += 1;
                if c["state"].as_str() == Some("running") {
                    containers_running += 1;
                }
            }
        }
        svc_map.insert(svc_name.clone(), svc_status);
    }

    Ok(Json(json!({
        "vm_id": vm_id,
        "label": label,
        "provider_state": probe.provider_state,
        "ssh": probe.ssh,
        "ping": probe.ping,
        "health": probe.health,
        "services": svc_map,
        "resources": resources,
        "summary": {
            "services_up": services_up,
            "services_down": services_down,
            "services_partial": services_partial,
            "containers_running": containers_running,
            "containers_total": containers_total,
        }
    })))
}
